### PR TITLE
Bugsnag ignore all action dispatch rescued errors

### PR DESCRIPTION
**Pull request by Martouta**
Closes [THREESCALE-4502](https://issues.redhat.com/browse/THREESCALE-4502)

With our middleware, it rescues but it pops up in Bugsnag anyway if we do not ignore these errors.
Without our middleware, it returns 4XX status code and it logs the errors even if we ignore these errors to not being reported to Bugsnag.

How to reproduce the exact requests and responses in Preview shown in https://issues.redhat.com/browse/THREESCALE-4502?focusedCommentId=14140999&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-14140999

Now ignoring these classes:
```ruby
["ActionController::RoutingError",
 "AbstractController::ActionNotFound",
 "ActionController::MethodNotAllowed",
 "ActionController::UnknownHttpMethod",
 "ActionController::NotImplemented",
 "ActionController::UnknownFormat",
 "ActionController::InvalidAuthenticityToken",
 "ActionController::InvalidCrossOriginRequest",
 "ActionDispatch::ParamsParser::ParseError",
 "ActionController::BadRequest",
 "ActionController::ParameterMissing",
 "Rack::QueryParser::ParameterTypeError",
 "Rack::QueryParser::InvalidParameterError",
 "ActiveRecord::RecordNotFound",
 "ActiveRecord::StaleObjectError",
 "ActiveRecord::RecordInvalid",
 "ActiveRecord::RecordNotSaved",
 "Roar::Rails::ControllerAdditions::UnsupportedMediaType",
 "WebHookWorker::ClientError"]
```

### CODE REVIEW

File: lib/three_scale/middleware/handle_bad_encoding_error.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+      end
+
+      attr_reader :app
+
+      def call(env)
+        # Rack::Request.new(env).params
+        # Rack::Utils.parse_nested_query(env['QUERY_STRING'].to_s)
+        app.call(env)
+      rescue Rack::QueryParser::InvalidParameterError => error
+        status = 440
```
Comment (Martouta) [inline lib/three_scale/middleware/handle_bad_encoding_error.rb:18]: I know that it should be 400 instead, but i am testing if it reaches here 😄

**Comment by Martouta** [comment]
### Filters
```
curl -v --request POST \
--header 'Accept: */*' \
--header 'Accept-Charset: iso-8859-1,utf-8;q=0.9,*;q=0.1' \
--header 'Accept-Language: en' \
--header 'Content-Type: application/x-www-form-urlencoded' \
-d '%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216boot%u002eini' \
'https://mnoya.preview01.3scale.net/admin/account/password'
```

```
curl -v --request POST \
--header 'Accept: */*' \
--header 'Accept-Charset: iso-8859-1,utf-8;q=0.9,*;q=0.1' \
--header 'Accept-Language: en' \
--header 'Content-Type: application/json' \
-d '{"foo" : "bar", ' \
'https://mnoya.preview01.3scale.net/admin/account/password'
```

```
curl -g -v --request GET \
  --header 'Accept: */*' \
  --header 'Accept-Charset: iso-8859-1,*,utf-8' \
  --header 'Accept-Language: en' \
  --header 'Cache-Control: no-cache' \
  --header 'Pragma: no-cache' \
  'https://mnoya.preview01.3scale.net/admin/account/password?_SERVER[]=&_SERVER[REMOTE_ADDR]=%27.system(%27id%27).exit().%27&option=wrapper&module[module]=1'
```

### Filtered before
```
curl -v "https://mnoya-admin.preview01.3scale.net/p/login?-1%'%20and%20(select%201%20from(select%20count(*)"
```

```
curl -v --request POST \
  --header 'Accept: */*' \
  --header 'Accept-Charset: iso-8859-1,utf-8;q=0.9,*;q=0.1' \
  --header 'Accept-Language: en' \
  --header 'Content-Type: application/x-www-form-urlencoded' \
  'https://mnoya.preview01.3scale.net/admin/account/password?%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216boot%u002eini'
```

**Comment by Martouta** [comment]
Interesting stuff 😄 

```
➜  deploy git:(master) curl -v --request POST \
--header 'Accept: */*' \
--header 'Accept-Charset: iso-8859-1,utf-8;q=0.9,*;q=0.1' \
--header 'Accept-Language: en' \
--header 'Content-Type: application/x-www-form-urlencoded' \
-d '%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216boot%u002eini' \
'https://mnoya.preview01.3scale.net/admin/account/password'
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 184.169.129.93...
* TCP_NODELAY set
* Connected to mnoya.preview01.3scale.net (184.169.129.93) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server did not agree to a protocol
* Server certificate:
*  subject: CN=*.preview01.3scale.net
*  start date: Sep 25 00:00:00 2019 GMT
*  expire date: Oct 25 12:00:00 2020 GMT
*  subjectAltName: host "mnoya.preview01.3scale.net" matched cert's "*.preview01.3scale.net"
*  issuer: C=US; O=Amazon; OU=Server CA 1B; CN=Amazon
*  SSL certificate verify ok.
> POST /admin/account/password HTTP/1.1
> Host: mnoya.preview01.3scale.net
> User-Agent: curl/7.64.1
> Accept: */*
> Accept-Charset: iso-8859-1,utf-8;q=0.9,*;q=0.1
> Accept-Language: en
> Content-Type: application/x-www-form-urlencoded
> Content-Length: 175
> 
* upload completely sent off: 175 out of 175 bytes
< HTTP/1.1 400 Bad Request
< Content-Type: text/html; charset=utf-8
< Date: Sat, 06 Jun 2020 17:10:06 GMT
< Server: openresty/1.5.12.1
< Strict-Transport-Security: max-age=15552000
< Vary: Accept-Encoding
< X-Content-Type-Options: nosniff
< X-Frame-Options: DENY
< X-Request-Id: 25f80ad2-e2c3-4e05-bb19-84532fd326e8
< X-Runtime: 0.016431
< X-Served-By: mt01
< X-XSS-Protection: 1; mode=block
< Content-Length: 0
< Connection: keep-alive
< 
* Connection #0 to host mnoya.preview01.3scale.net left intact
* Closing connection 0
➜  deploy git:(master) curl -v --request POST \
--header 'Accept: */*' \
--header 'Accept-Charset: iso-8859-1,utf-8;q=0.9,*;q=0.1' \
--header 'Accept-Language: en' \
--header 'Content-Type: application/json' \
-d '{"foo" : "bar", ' \
'https://mnoya.preview01.3scale.net/admin/account/password'
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 184.72.58.149...
* TCP_NODELAY set
* Connected to mnoya.preview01.3scale.net (184.72.58.149) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server did not agree to a protocol
* Server certificate:
*  subject: CN=*.preview01.3scale.net
*  start date: Sep 25 00:00:00 2019 GMT
*  expire date: Oct 25 12:00:00 2020 GMT
*  subjectAltName: host "mnoya.preview01.3scale.net" matched cert's "*.preview01.3scale.net"
*  issuer: C=US; O=Amazon; OU=Server CA 1B; CN=Amazon
*  SSL certificate verify ok.
> POST /admin/account/password HTTP/1.1
> Host: mnoya.preview01.3scale.net
> User-Agent: curl/7.64.1
> Accept: */*
> Accept-Charset: iso-8859-1,utf-8;q=0.9,*;q=0.1
> Accept-Language: en
> Content-Type: application/json
> Content-Length: 16
> 
* upload completely sent off: 16 out of 16 bytes
< HTTP/1.1 400 Bad Request
< Content-Type: text/html; charset=utf-8
< Date: Sat, 06 Jun 2020 17:11:06 GMT
< Server: openresty/1.5.12.1
< Strict-Transport-Security: max-age=15552000
< Vary: Accept-Encoding
< X-Content-Type-Options: nosniff
< X-Frame-Options: DENY
< X-Request-Id: cc71b024-c95c-450b-a913-df37497b1201
< X-Runtime: 0.712821
< X-Served-By: mt01
< X-XSS-Protection: 1; mode=block
< Content-Length: 0
< Connection: keep-alive
< 
* Connection #0 to host mnoya.preview01.3scale.net left intact
* Closing connection 0
➜  deploy git:(master) curl -g -v --request GET \
  --header 'Accept: */*' \
  --header 'Accept-Charset: iso-8859-1,*,utf-8' \
  --header 'Accept-Language: en' \
  --header 'Cache-Control: no-cache' \
  --header 'Pragma: no-cache' \
  'https://mnoya.preview01.3scale.net/admin/account/password?_SERVER[]=&_SERVER[REMOTE_ADDR]=%27.system(%27id%27).exit().%27&option=wrapper&module[module]=1'
Note: Unnecessary use of -X or --request, GET is already inferred.
*   Trying 50.18.117.80...
* TCP_NODELAY set
* Connected to mnoya.preview01.3scale.net (50.18.117.80) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server did not agree to a protocol
* Server certificate:
*  subject: CN=*.preview01.3scale.net
*  start date: Sep 25 00:00:00 2019 GMT
*  expire date: Oct 25 12:00:00 2020 GMT
*  subjectAltName: host "mnoya.preview01.3scale.net" matched cert's "*.preview01.3scale.net"
*  issuer: C=US; O=Amazon; OU=Server CA 1B; CN=Amazon
*  SSL certificate verify ok.
> GET /admin/account/password?_SERVER[]=&_SERVER[REMOTE_ADDR]=%27.system(%27id%27).exit().%27&option=wrapper&module[module]=1 HTTP/1.1
> Host: mnoya.preview01.3scale.net
> User-Agent: curl/7.64.1
> Accept: */*
> Accept-Charset: iso-8859-1,*,utf-8
> Accept-Language: en
> Cache-Control: no-cache
> Pragma: no-cache
> 
< HTTP/1.1 400 Bad Request
< Content-Type: text/html; charset=utf-8
< Date: Sat, 06 Jun 2020 17:11:17 GMT
< Server: openresty/1.5.12.1
< Strict-Transport-Security: max-age=15552000
< Vary: Accept-Encoding
< X-Content-Type-Options: nosniff
< X-Frame-Options: DENY
< X-Request-Id: 04c0da07-6326-4ee8-81ef-e2a9c40ad988
< X-Runtime: 0.005338
< X-Served-By: mt01
< X-XSS-Protection: 1; mode=block
< Content-Length: 0
< Connection: keep-alive
< 
* Connection #0 to host mnoya.preview01.3scale.net left intact
* Closing connection 0
➜  deploy git:(master) curl -v "https://mnoya-admin.preview01.3scale.net/p/login?-1%'%20and%20(select%201%20from(select%20count(*)"
*   Trying 50.18.105.24...
* TCP_NODELAY set
* Connected to mnoya-admin.preview01.3scale.net (50.18.105.24) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server did not agree to a protocol
* Server certificate:
*  subject: CN=*.preview01.3scale.net
*  start date: Sep 25 00:00:00 2019 GMT
*  expire date: Oct 25 12:00:00 2020 GMT
*  subjectAltName: host "mnoya-admin.preview01.3scale.net" matched cert's "*.preview01.3scale.net"
*  issuer: C=US; O=Amazon; OU=Server CA 1B; CN=Amazon
*  SSL certificate verify ok.
> GET /p/login?-1%'%20and%20(select%201%20from(select%20count(*) HTTP/1.1
> Host: mnoya-admin.preview01.3scale.net
> User-Agent: curl/7.64.1
> Accept: */*
> 
< HTTP/1.1 400 Bad Request
< Date: Sat, 06 Jun 2020 17:11:31 GMT
< Server: openresty/1.5.12.1
< Content-Length: 0
< Connection: keep-alive
< 
* Connection #0 to host mnoya-admin.preview01.3scale.net left intact
* Closing connection 0
➜  deploy git:(master) curl -v --request POST \
  --header 'Accept: */*' \
  --header 'Accept-Charset: iso-8859-1,utf-8;q=0.9,*;q=0.1' \
  --header 'Accept-Language: en' \
  --header 'Content-Type: application/x-www-form-urlencoded' \
  'https://mnoya.preview01.3scale.net/admin/account/password?%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216boot%u002eini'
*   Trying 50.18.117.80...
* TCP_NODELAY set
* Connected to mnoya.preview01.3scale.net (50.18.117.80) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server did not agree to a protocol
* Server certificate:
*  subject: CN=*.preview01.3scale.net
*  start date: Sep 25 00:00:00 2019 GMT
*  expire date: Oct 25 12:00:00 2020 GMT
*  subjectAltName: host "mnoya.preview01.3scale.net" matched cert's "*.preview01.3scale.net"
*  issuer: C=US; O=Amazon; OU=Server CA 1B; CN=Amazon
*  SSL certificate verify ok.
> POST /admin/account/password?%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216%u002e%u002e%u2216boot%u002eini HTTP/1.1
> Host: mnoya.preview01.3scale.net
> User-Agent: curl/7.64.1
> Accept: */*
> Accept-Charset: iso-8859-1,utf-8;q=0.9,*;q=0.1
> Accept-Language: en
> Content-Type: application/x-www-form-urlencoded
> 
< HTTP/1.1 400 Bad Request
< Date: Sat, 06 Jun 2020 17:11:40 GMT
< Server: openresty/1.5.12.1
< Content-Length: 0
< Connection: keep-alive
< 
* Connection #0 to host mnoya.preview01.3scale.net left intact
* Closing connection 0
➜  deploy git:(master) curl -v --request --output GET \
  --header 'X-Real-Ip: 34.200.223.28' \
  --header 'X-Request-Start: t=1591264843.359' \
  --header 'Accept: */*' \
  --header 'Accept-Encoding: gzip, deflate, br' \
  --header 'Api-Key: 77948fafb563eb1dd6dd2466234d0591' \
  --header 'Content-Type: application/json' \
  --header 'Postman-Token: 227856cf-a5f7-4e41-831a-47b0ac5be35d' \
  --header 'User-Agent: PostmanRuntime/7.25.0' \
  --header 'Content-Type: application/json' \
  --header 'X-Request-Id: ac1120ea550f0e3c2eee8fb5562b6d4b' \
  --header 'Content-Length: 27' \
  'https://mnoya.preview01.3scale.net/admin/account'
* Could not resolve host: GET
* Closing connection 0
curl: (6) Could not resolve host: GET
*   Trying 50.18.117.80...
* TCP_NODELAY set
* Connected to mnoya.preview01.3scale.net (50.18.117.80) port 443 (#1)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server did not agree to a protocol
* Server certificate:
*  subject: CN=*.preview01.3scale.net
*  start date: Sep 25 00:00:00 2019 GMT
*  expire date: Oct 25 12:00:00 2020 GMT
*  subjectAltName: host "mnoya.preview01.3scale.net" matched cert's "*.preview01.3scale.net"
*  issuer: C=US; O=Amazon; OU=Server CA 1B; CN=Amazon
*  SSL certificate verify ok.
> --output /admin/account HTTP/1.1
> Host: mnoya.preview01.3scale.net
> X-Real-Ip: 34.200.223.28
> X-Request-Start: t=1591264843.359
> Accept: */*
> Accept-Encoding: gzip, deflate, br
> Api-Key: 77948fafb563eb1dd6dd2466234d0591
> Content-Type: application/json
> Postman-Token: 227856cf-a5f7-4e41-831a-47b0ac5be35d
> User-Agent: PostmanRuntime/7.25.0
> Content-Type: application/json
> X-Request-Id: ac1120ea550f0e3c2eee8fb5562b6d4b
> Content-Length: 27
> 
< HTTP/1.1 400 Bad Request
< Content-Type: text/html; charset=UTF-8
< Date: Sat, 06 Jun 2020 17:11:46 GMT
< Server: openresty/1.5.12.1
< Content-Length: 179
< Connection: keep-alive
< 
<html>
<head><title>400 Bad Request</title></head>
<body bgcolor="white">
<center><h1>400 Bad Request</h1></center>
<hr><center>openresty/1.5.12.1</center>
</body>
</html>
* Connection #1 to host mnoya.preview01.3scale.net left intact
* Closing connection 1
```

**Comment by Martouta** [comment]
Hery told me that I might be able to ignore from Bugsnag all the classes from `Rails.application.action_dispatch.rescue_responses`

File: config/initializers/bugsnag.rb [hunk 2]
```diff
@@ -1,10 +1,17 @@
...
 Bugsnag.configure do |config|
   config.api_key = Rails.configuration.three_scale.bugsnag_api_key
   config.app_version = System::Deploy.info.revision
   config.notify_release_stages = %w[production]
   stages = Rails.configuration.three_scale.error_reporting_stages
   config.notify_release_stages = stages if stages.present?
 
-  # when WebHooks fails because the remote is down or similar - we don't mind
-  config.ignore_classes << ->(error) { WebHookWorker::ClientError === error }
+  ignore_error_names = ActionDispatch::ExceptionWrapper.rescue_responses.each_with_object([]) do |(key, value), classes|
```
Comment (hallelujah) [inline config/initializers/bugsnag.rb:10]: Why not just add all of them directly, because rails considers them to be rescued anyway

**Pull request merged**